python/python/tests/test_optimize.py

```diff
@@ -0,0 +1,87 @@
+#  Copyright (c) 2023. Lance Developers
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import pickle
+import re
+from pathlib import Path
+
+import lance
+import pyarrow as pa
+from lance.lance import Compaction
+from lance.optimize import CompactionPlan, CompactionTask, RewriteResult
+
+
+def test_dataset_optimize(tmp_path: Path):
+    base_dir = tmp_path / "dataset"
+    data = pa.table({"a": range(1000), "b": range(1000)})
+
+    dataset = lance.write_dataset(data, base_dir, max_rows_per_file=100)
+    assert dataset.version == 1
+    assert len(dataset.get_fragments()) == 10
+
+    metrics = dataset.optimize.compact_files(
+        target_rows_per_fragment=1000,
+        materialize_deletions=False,
+        num_threads=1,
+    )
+
+    assert metrics.fragments_removed == 10
+    assert metrics.fragments_added == 1
+    assert metrics.files_removed == 10
+    assert metrics.files_added == 1
+
+    assert dataset.version == 2
+
+
+def test_dataset_distributed_optimize(tmp_path: Path):
+    base_dir = tmp_path / "dataset"
+    data = pa.table({"a": range(800), "b": range(800)})
+
+    dataset = lance.write_dataset(data, base_dir, max_rows_per_file=200)
+    fragments = dataset.get_fragments()
+    assert len(fragments) == 4
+
+    plan = Compaction.plan(
+        dataset, options=dict(target_rows_per_fragment=400, num_threads=None)
+    )
+    assert plan.read_version == 1
+    assert plan.num_tasks() == 2
+    assert plan.tasks[0].fragments == [frag.metadata for frag in fragments[0:2]]
+    assert plan.tasks[1].fragments == [frag.metadata for frag in fragments[2:4]]
+    assert repr(plan) == "CompactionPlan(read_version=1, tasks=<2 compaction tasks>)"
+    # Plan can be pickled
+    assert pickle.loads(pickle.dumps(plan)) == plan
+    assert isinstance(plan, CompactionPlan)
+
+    pickled_task = pickle.dumps(plan.tasks[0])
+    task = pickle.loads(pickled_task)
+    assert task == plan.tasks[0]
+    assert isinstance(task, CompactionTask)
```

Comment:
```suggestion
    task = pickle.loads(pickled_task)
    assert isinstance(task, CompactionTask)
    assert task == plan.tasks[0]
```
Again, not entirely certain what this is testing, but it seems more useful to verify the pickle-round-tripped object